src/py/idom/codec/transpile.py

```diff
@@ -0,0 +1,161 @@
+from io import StringIO
+from typing import Tuple, Optional, Sequence, List
+
+# we need to import these because HTMLParser makes tags and attributes lowercase by
+# default which we don't want - as a result we reimplement `HTMLParser.parse_starttag`
+# without that behavior.
+from html.parser import tagfind_tolerant, attrfind_tolerant, unescape  # type: ignore
+
+from .utils import split_fstr_style_exprs, transform_string
+
+
+def transpile_html_templates(text: str) -> str:
+    return transform_string(text, _transpile_html_templates, _TemplateTranspiler())
+
+
+def _transpile_html_templates(
+    text: str, index: int, transpiler: "_TemplateTranspiler"
+) -> Tuple[int, Optional[str]]:
+    if text[index : index + 4] != "html":
+        return index + 1, None
+
+    index += 4
+
+    for str_char in ('"""', '"', "'''", "'"):
+        # characters open a string
+        if text[index : index + len(str_char)] == str_char:
+            break
+    else:
+        return index + 1, None
+
+    for forward_index in range(index + 1, len(text)):
+        if text[forward_index : forward_index + len(str_char)] == str_char:
+            template = text[index + len(str_char) : forward_index]
+            py_code = transpiler.transpile(template.strip())
+            return forward_index + len(str_char), py_code
+
+    return forward_index + len(str_char), None
+
+
+class _TemplateTranspiler(HTMLParser):
+    def transpile(self, text: str) -> str:
+        self._feed(text)
+        code = self._read_code()
+        self.reset()
+        return code
+
+    def feed(self, text: str) -> None:
+        raise NotImplementedError()
+
+    def _feed(self, text: str) -> None:
+        return super().feed(self._load_exprs(text))
+
+    def _read_code(self) -> str:
+        self._code.seek(0)
+        value = self._code.read()
+        self._code.seek(self._cursor)
+        return value[:-2]
+
+    def reset(self) -> None:
+        super().reset()
+        self._exprs: List[str] = []
+        self._cursor = 0
+        self._code = StringIO()
+
+    def handle_starttag(
+        self, tag: str, attrs: Sequence[Tuple[str, Optional[str]]]
+    ) -> None:
+        self._write(f"idom.vdom({self._substitute_exprs(tag)}, " + "{")
+        subbed_attrs = [
+            (
+                self._substitute_exprs(k),
+                self._substitute_exprs("true" if v is None else v),
+            )
+            for k, v in attrs
+        ]
+        self._write(", ".join("%s: %s" % item for item in subbed_attrs))
+        self._write("}, [")
+
+    def handle_data(self, data: str) -> None:
+        self._write(self._substitute_exprs(data) + ", ")
+
+    def handle_endtag(self, tag: str) -> None:
+        self._write("]), ")
+
+    def handle_startendtag(self, tag: str, attrs: Sequence[Tuple[str, str]]) -> None:
+        self.handle_starttag(tag, attrs)
+        self.handle_endtag(tag)
+
+    def parse_starttag(self, i: int) -> int:
+        self.__starttag_text = None
```

Comment:
Add note that this was copy-pasted from:
https://github.com/python/cpython/blob/master/Lib/html/parser.py